### PR TITLE
feat: add LoRA (Low-Rank Adaptation) for parameter-efficient finetuning

### PR DESCRIPTION
**Pull request by gabo515**
## Summary

Add LoRA (Low-Rank Adaptation) support to Savanna, enabling parameter-efficient finetuning of StripedHyena/Evo models by training low-rank adapter matrices while keeping the base model frozen.

- Wraps `ColumnParallelLinear`, `RowParallelLinear`, and `TELinear` (FP8) with LoRA adapters
- YAML-configurable: rank, alpha, dropout, target modules, separate LR, freeze control
- Save/load LoRA weights separately per MP rank, or merged into base weights
- Checkpoint key remapping for loading base checkpoints into LoRA-wrapped models
- Compatible with activation checkpointing (`_ensure_requires_grad` fix for frozen base + reentrant checkpoint)
- Config validation catches common misconfigurations (bad rank, float rank, string target_modules)

### Goal
Enable parameter-efficient finetuning of Savanna models by adding LoRA adapters to configurable linear layers, while keeping the base model frozen.

### Supported layer types
| Layer | LoRA wrapper | Notes |
|---|---|---|
| `ColumnParallelLinear` | `LoRAColumnParallelLinear` | A: full input, B: partitioned output |
| `RowParallelLinear` | `LoRARowParallelLinear` | A: partitioned input, B: full output |
| `TELinear` (FP8) | `LoRATELinear` | bf16 LoRA params on top of FP8 base |

### Configuration
```yaml
lora:
  enabled: true
  r: 16                    # rank
  alpha: 32.0              # scaling = alpha / r
  dropout: 0.05
  target_modules:          # leaf module names to wrap
    - "dense_projection"   # QKV (ColumnParallel)
    - "dense"              # output proj (RowParallel)
    - "w1"                 # GLU gate (ColumnParallel)
    - "w2"                 # GLU up-proj (ColumnParallel)
    - "w3"                 # GLU down-proj (RowParallel)
  freeze_base_model: true
  modules_to_save: []      # additional modules to keep trainable
  save_merged: false       # merge LoRA into base on save
  lora_lr: 0.0003          # separate LR for LoRA params
  lora_weight_decay: 0.0
```

### Key design decisions
- **LoRA params in bf16** even when base uses FP8 — avoids quantization noise in low-rank adapters
- **Checkpoint key remapping** (`_make_lora_custom_load_fn`) — transparently loads base checkpoints into LoRA-wrapped models by inserting `base_layer.` prefix
- **`_ensure_requires_grad`** — fixes interaction between frozen base model and DeepSpeed's reentrant activation checkpointing, which requires at least one input tensor with `requires_grad=True`
- **Merge/unmerge** saves original weight for exact restoration (avoids bf16 precision loss from add-then-subtract)
- **Async save guard** — forces blocking save when `save_merged=True` to prevent race condition with unmerge

### Out of scope
- Pipeline parallelism with LoRA
- QLoRA / quantized base weights
- LoRA for non-linear layers (embeddings, norms)

## Files changed

| File | Change |
|---|---|
| `savanna/lora.py` | **New.** LoRA wrapper classes, apply/merge/unmerge, state dict helpers |
| `savanna/arguments/lora_config.py` | **New.** `GlobalConfigLoRA` dataclass with validation |
| `tests/unit/test_lora.py` | **New.** 64 unit tests |
| `savanna/arguments/arguments.py` | Register `GlobalConfigLoRA` in `BASE_CLASSES` |
| `savanna/checkpointing.py` | LoRA save/load, key remapping, async save race fix |
| `savanna/model/utils.py` | LoRA param groups, `_ensure_requires_grad` for activation ckpt |
| `savanna/training.py` | Inject LoRA in `get_model()`, freeze base, finetune load path |

## Test plan

- [x] 64 unit tests covering all three wrapper classes, merge/unmerge, state dict, config validation, `_ensure_requires_grad`, backward passes
- [x] Lint clean (`ruff check` passes on all changed files)
- [x] Smoke tested: single-GPU LoRA finetuning of Evo2-7B with FP8/TELinear, loss decreasing, 30% MFU

### CODE REVIEW

**Comment by Zymrael** [comment]
Thanks for the contribution. Do you have some example test loss curves to share before we merge?

**Comment by gabo515** [comment]
image

If you would prefer some other ones or a rank sweep, I'd be happy to run that

**Comment by Zymrael** [comment]
Thanks! Merging
